datafusion/expr/src/expr_schema.rs

```diff
@@ -112,7 +112,22 @@ impl ExprSchemable for Expr {
             Expr::OuterReferenceColumn(ty, _) => Ok(ty.clone()),
             Expr::ScalarVariable(ty, _) => Ok(ty.clone()),
             Expr::Literal(l) => Ok(l.data_type()),
-            Expr::Case(case) => case.when_then_expr[0].1.get_type(schema),
+            // Expr::Case(case) => case.when_then_expr[0].1.get_type(schema),
+            Expr::Case(case) => {
+                let then_type = case.when_then_expr[0].1.get_type(schema)?;
+                if !then_type.is_null() {
+                    return Ok(then_type);
+                }
+
+                let else_type = case
+                    .else_expr
+                    .as_ref()
+                    .map_or(Ok(DataType::Null), |e| e.get_type(schema))?;
+                if !else_type.is_null() {
+                    return Ok(else_type);
+                }
+                Ok(DataType::Int64)
```

Comment:
Although I think returning `DataType::Int64` can make it work fine, because after type coercion it always returns `then_type`. But it is still strange, the type of `CASE data WHEN 1 THEN NULL WHEN 2 THEN 'FOO' ELSE NULL END` is  `DataType::Int64`.

How about returning the first non-NULL type among the `then`  and `else` expressions?

Reply:
Current implementation will return the first non-NULL type among the then and else expressions.

``` shell
> select arrow_typeof(CASE data WHEN 1 THEN NULL WHEN 2 THEN 'FOO' ELSE NULL END )from example;
+------------------------------------------------------------------------------------------------------+
| arrow_typeof(CASE example.data WHEN Int64(1) THEN NULL WHEN Int64(2) THEN Utf8("FOO") ELSE NULL END) |
+------------------------------------------------------------------------------------------------------+
| Utf8                                                                                                 |
| Utf8                                                                                                 |
| Utf8                                                                                                 |
| Utf8                                                                                                 |
+------------------------------------------------------------------------------------------------------+
```

Reply:
This is because `type_coercion` will convert their types to `Utf8`.
```sh
| logical_plan after type_coercion                           | Projection: arrow_typeof(CASE CAST(example.data AS Int64) WHEN Int64(1) THEN CAST(NULL AS Utf8) WHEN Int64(2) THEN Utf8("FOO") ELSE CAST(NULL AS Utf8) END)                                                                                                                                                                                    |
|                                                            |   TableScan: example
```
You can check this [test](https://github.com/jonahgao/datafusion/blob/d9ff0cafd01614fe850ef4c12ba67761613336e8/datafusion/expr/src/expr_schema.rs#L677). It does not perform type coercion.

Reply:
done